src/System.Data.Common/src/System/Data/Common/DbDataReaderExtensions.cs

```diff
@@ -2,22 +2,97 @@
+                columnSchema.Add(dbColumn);
+            }
+            System.Collections.ObjectModel.ReadOnlyCollection<DbColumn> readOnlyColumnSchema = new System.Collections.ObjectModel.ReadOnlyCollection<DbColumn>(columnSchema);
+            return readOnlyColumnSchema;
+        }
+    }
+
+    internal sealed class DataRowDbColumn : DbColumn
+    {
+        private DataColumnCollection _schemaColumns;
+        private DataRow _schemaRow;
+
+        public DataRowDbColumn(DataRow readerSchemaRow, DataColumnCollection readerSchemaColumns)
+        {
+            _schemaRow = readerSchemaRow;
+            _schemaColumns = readerSchemaColumns;
+            populateFields();
+        }
+
+        private void populateFields()
+        {
+            AllowDBNull = GetDbColumnValue<bool?>(SchemaTableColumn.AllowDBNull);
+            BaseCatalogName = GetDbColumnValue<string>(SchemaTableOptionalColumn.BaseCatalogName);
+            BaseColumnName = GetDbColumnValue<string>(SchemaTableColumn.BaseColumnName);
+            BaseSchemaName = GetDbColumnValue<string>(SchemaTableColumn.BaseSchemaName);
+            BaseServerName = GetDbColumnValue<string>(SchemaTableOptionalColumn.BaseServerName);
+            BaseTableName = GetDbColumnValue<string>(SchemaTableColumn.BaseTableName);
+            ColumnName = GetDbColumnValue<string>(SchemaTableColumn.ColumnName);
+            ColumnOrdinal = GetDbColumnValue<int?>(SchemaTableColumn.ColumnOrdinal);
+            ColumnSize = GetDbColumnValue<int?>(SchemaTableColumn.ColumnSize);
+            IsAliased = GetDbColumnValue<bool?>(SchemaTableColumn.IsAliased);
+            IsAutoIncrement = GetDbColumnValue<bool?>(SchemaTableOptionalColumn.IsAutoIncrement);
+            IsExpression = GetDbColumnValue<bool>(SchemaTableColumn.IsExpression);
+            IsHidden = GetDbColumnValue<bool?>(SchemaTableOptionalColumn.IsHidden);
+            IsIdentity = GetDbColumnValue<bool?>("IsIdentity");
+            IsKey = GetDbColumnValue<bool?>(SchemaTableColumn.IsKey);
+            IsLong = GetDbColumnValue<bool?>(SchemaTableColumn.IsLong);
+            IsReadOnly = GetDbColumnValue<bool?>(SchemaTableOptionalColumn.IsReadOnly);
+            IsUnique = GetDbColumnValue<bool?>(SchemaTableColumn.IsUnique);
+            NumericPrecision = GetDbColumnValue<int?>(SchemaTableColumn.NumericPrecision);
+            NumericScale = GetDbColumnValue<int?>(SchemaTableColumn.NumericScale);
+            UdtAssemblyQualifiedName = GetDbColumnValue<string>("UdtAssemblyQualifiedName");
+            DataType = GetDbColumnValue<Type>(SchemaTableColumn.DataType);
+            DataTypeName = GetDbColumnValue<string>("DataTypeName");
+        }
+
+        private T GetDbColumnValue<T>(string columnName)
+        {
+            if (!_schemaColumns.Contains(columnName))
+            {
+                return default(T);
+            }
+
+            object schemaObject = _schemaRow[columnName];
```

Comment:
nit: simplify with conditional operator and pattern matching (whole method could probably be expression-bodied)

Reply:
Do you mean go full c#8 on it?

Reply:
I had a go. You can express it as an expression bodies method but the readability suffers a great deal. It's simpler and clearer now I think, have a look.